src/hashtree.erl

```diff
@@ -355,38 +369,68 @@ update_tree(State) ->
     update_perform(State3).
 
 -spec update_perform(hashtree()) -> hashtree().
-update_perform(State2=#state{dirty_segments=Dirty, segments=NumSegments}) ->
-    %% Segments = gb_sets:to_list(Dirty),
-    Segments = bitarray_to_list(Dirty),
+update_perform(State=#state{dirty_segments=Dirty, segments=NumSegments}) ->
+    NextRebuild = State#state.next_rebuild,
+    Segments = case NextRebuild of
+                   full ->
+                       ?ALL_SEGMENTS;
+                   incremental ->
+                       %% gb_sets:to_list(Dirty),
+                       bitarray_to_list(Dirty)
+               end,
+    State2 = maybe_clear_buckets(Segments, State),
     State3 = update_tree(Segments, State2),
-    %% State3#state{dirty_segments=gb_sets:new()}.
-    State3#state{dirty_segments=bitarray_new(NumSegments)}.
+    %% State2#state{dirty_segments=gb_sets:new()}
+    State3#state{dirty_segments=bitarray_new(NumSegments),
+                 next_rebuild=incremental}.
+
+maybe_clear_buckets(?ALL_SEGMENTS, State) ->
+    Segments = lists:seq(0, State#state.segments div State#state.width),
+    %% Limit the clear to the disk buckets, then dump the whole tree
+    State2 = clear_bucket(State#state.mem_levels, State#state.levels, Segments, State),
+    State2#state{tree = dict:new()};
+maybe_clear_buckets(_Segments, State) ->
+    State.
+
+
+clear_bucket(MinLevel, Level, _Segments, State) when Level =< MinLevel ->
+    State;
+clear_bucket(MinLevel, Level, Segments, State = #state{width = Width}) ->
+    %% ignoring del_bucket state return
+    State2 = lists:foldl(fun(Segment, State1) ->
+                                 del_bucket(Level, Segment, State1)
```

Comment:
del_memory_bucket isn't actually used any more, could simplify and remove del_bucket/del_memory_bucket and just call ```del_disk_bucket``` in the clean. Can't think of any cases where we would need to delete memory buckets individually.